### PR TITLE
LUCENE-10354: Clarify contract of codec APIs with missing/disabled fields.

### PR DESCRIPTION
**Pull request by jpountz**
See https://issues.apache.org/jira/browse/LUCENE-10354

### CODE REVIEW

**Review by jtibshirani — APPROVED** [review]
Nice! Now I won't ask you about the behavior on every code review :)

File: lucene/core/src/java/org/apache/lucene/codecs/FieldsProducer.java [hunk 1]
```diff
@@ -42,6 +45,14 @@ protected FieldsProducer() {}
    */
   public abstract void checkIntegrity() throws IOException;
 
+  /**
+   * Get the {@link Terms} for this field. The behavior is undefined if the field doesn't have
```
[the remark this answers is not in the document]
Comment (jpountz) [reply lucene/core/src/java/org/apache/lucene/codecs/FieldsProducer.java]: This is the part of the change I was not comfortable with. I wanted to do the change on codec APIs while still giving the ability for users to pass any fields in user-facing APIs, unfortunately Fields is a user-facing API so I didn't want to change it so I thought of imposing a stricter contract on `FieldsProducer`. I reverted this part of the change for now as I'd like to think more about it.

There is another issue with this change: postings and points currently return null for ghost fields. I'll open a separate issue to ask if we want to make this more consistent with doc values or vectors which return empty instances for ghost fields (ie. fields that have the feature enabled on their FieldInfos but no values). So I reverted the comment that says that the return value is not null on points and postings.

[the remark this answers is not in the document]
Comment (jpountz) [reply lucene/core/src/java/org/apache/lucene/codecs/FieldsProducer.java]: I raised the question about ghost fields at https://issues.apache.org/jira/browse/LUCENE-10357.

[the remark this answers is not in the document]
Comment (jtibshirani) [reply lucene/core/src/java/org/apache/lucene/codecs/FieldsProducer.java]: Got it. The latest changes make sense to me.

**Pull request merged**